File: src/typescript/src/xiaoyi/core/config/builder.ts
/**
 * # Configuration Builder
 *
 * `builder` provides a fluent builder pattern for constructing configurations.
 *
 * Path: `xiaoyi.core.config.builder`
 *
 * @module core.config.builder
 * @brief Fluent configuration builder
 * @group Core
 * @since 0.1.0
 * @author Miruamel
 * @see core.config
 * @see core.config.source
 */
import { Config, ConfigSource } from "./config";

/**
 * Configuration builder for composing multiple sources.
 *
 * @brief Build configuration from multiple sources
 * @group Core
 * @since 0.1.0
 * @example
 * ```typescript
 * const config = await new ConfigBuilder()
 *   .addSource(new FileSource({ path: "./config.toml" }))
 *   .addSource(new EnvSource({ prefix: "XIAOYI_" }))
 *   .build();
 * ```
 */
export class ConfigBuilder {
  private sources: ConfigSource[] = [];

  /**
   * Add a configuration source.
   *
   * @param source - Configuration source to add
   * @returns this (for chaining)
   * @since 0.1.0
   */
  addSource(source: ConfigSource): this {
    this.sources.push(source);
    return this;
  }

  /**
   * Add multiple configuration sources.
   *
   * @param sources - Configuration sources to add
   * @returns this (for chaining)
   * @since 0.1.0
   */
  addSources(sources: ConfigSource[]): this {
    this.sources.push(...sources);
    return this;
  }

  /**
   * Build configuration by loading all sources in priority order.
   *
   * @returns Merged configuration
   * @since 0.1.0
   */
  async build(): Promise<Config> {
    // Sort by priority (lower first, so higher priority overrides)
    const sorted = [...this.sources].sort((a, b) => a.priority - b.priority);

    const data: Record<string, unknown> = {};
    const sourceNames: string[] = [];

    for (const source of sorted) {
      const loaded = await source.load();
      Object.assign(data, loaded);
      sourceNames.push(source.name);
    }

    return { data, sources: sourceNames };
  }
}

File: src/python/xiaoyi/core/config/builder.py
"""
# Configuration Builder

`builder` provides a fluent builder pattern for constructing configurations.

Path: `xiaoyi.core.config.builder`

@module xiaoyi.core.config.builder
@brief Fluent configuration builder
@group Core
@since 0.1.0
@author Miruamel
@see xiaoyi.core.config
@see xiaoyi.core.config.source
"""

from typing import List
from .config import Config, ConfigSource


class ConfigBuilder:
    """
    Configuration builder for composing multiple sources.

    @brief Build configuration from multiple sources
    @group Core
    @since 0.1.0
    @example
    ```python
    config = await (ConfigBuilder()
        .add_source(FileSource(path="./config.toml"))
        .add_source(EnvSource(prefix="XIAOYI_"))
        .build())
    ```
    """

    def __init__(self):
        self._sources: List[ConfigSource] = []

    def add_source(self, source: ConfigSource) -> "ConfigBuilder":
        """
        Add a configuration source.

        @param source Configuration source to add
        @return this (for chaining)
        @since 0.1.0
        """
        self._sources.append(source)
        return self

    def add_sources(self, sources: List[ConfigSource]) -> "ConfigBuilder":
        """
        Add multiple configuration sources.

        @param sources Configuration sources to add
        @return this (for chaining)
        @since 0.1.0
        """
        self._sources.extend(sources)
        return self

    async def build(self) -> Config:
        """
        Build configuration by loading all sources in priority order.

        @return Merged configuration
        @since 0.1.0
        """
        # Sort by priority (lower first, so higher priority overrides)
        sorted_sources = sorted(self._sources, key=lambda s: s.priority)

        data: Dict[str, Any] = {}
        source_names: List[str] = []

        for source in sorted_sources:
            loaded = await source.load()
            data.update(loaded)
            source_names.append(source.name)

        return Config(data=data, sources=source_names)


__all__ = ["ConfigBuilder"]